pkg/arvo/sys/zuse.hoon

```diff
@@ -2313,21 +2313,112 @@
     ::
     ::  %kale only talks to %ames and itself.
     ::
+    ++  block
+      =<  block
+      |%
+      +$  hash    @uxblockhash
+      +$  number  @udblocknumber
+      +$  id      [=hash =number]
+      +$  block   [=id =parent=hash]
+      --
+    ::
+    ::  Azimuth points form a groupoid, where the objects are all the
+    ::  possible values of +point and the arrows are the possible values
+    ::  of (list point-diff).  Composition of arrows is concatenation,
+    ::  and you can apply the diffs to a +point with +apply.
+    ::
+    ::  It's simplest to consider +point as the product of three
+    ::  groupoids, Rift, Keys, and Sponsor.  The objects of the product
+    ::  are the product of the objects of the underlying groupoids.  The
+    ::  arrows are a list of a sum of the diff types of the underlying
+    ::  groupoids.  Given an arrow=(list diff), you can project to the
+    ::  underlying arrows with +skim filtering on the head of each
+    ::  diff.
+    ::
+    ::  The identity element is ~.  Clearly, composing this with any +diff
+    ::  gives the original +diff.  Since this is a category, +compose must
+    ::  be associative (true, because concatenation is associative).  This
+    ::  is a groupoid, so we must further have that every +point-diff has an
+    ::  inverse.  These are given by the +inverse operation.
+    ::
     ++  point
-      $:  =rift
-          =life
-          keys=(map life [crypto-suite=@ud =pass])
-          sponsor=(unit @p)
-      ==
-    +$  point-diff
-      $%  [%changed-continuity =rift]
-          [%changed-keys =life crypto-suite=@ud =pass]
-          [%new-sponsor sponsor=(unit @p)]
-      ==
+      =<  point
+      |%
+      +$  point
+        $:  =rift
+            =life
+            keys=(map life [crypto-suite=@ud =pass])
+            sponsor=(unit @p)
+        ==
+      ::
+      +$  key-update  [=life crypto-suite=@ud =pass]
+      ::
+      ::  Invertible diffs
+      ::
+      +$  diffs  (list diff)
+      +$  diff
+        $%  [%rift from=rift to=rift]
+            [%keys from=key-update to=key-update]
+            [%spon from=(unit @p) to=(unit @p)]
+        ==
+      ::
+      ::  Non-invertible diffs
+      ::
+      +$  udiffs  (list [=ship =udiff])
+      +$  udiff
+        $:  =id:block
+        $%  [%rift =rift]
+            [%keys key-update]
+            [%spon sponsor=(unit @p)]
+            [%disavow ~]
+        ==  ==
+      ::
+      ++  inverse
+        |=  diffs=(list diff)
+        ^-  (list diff)
+        %-  flop
+        %+  turn  diffs
+        |=  =diff
+        ^-  ^diff
+        ?-  -.diff
+          %rift  [%rift &2 |2]:diff
+          %keys  [%keys &2 |2]:diff
+          %spon  [%spon &2 |2]:diff
```

Comment:
Is there a way to write this that's more common? This seems like tree addressing, but I haven't seen this notation before. (It's up in azimuth-tracker's `+parse-block` also, just noticed.)

Reply:
I'll use `+.diff`.